File: crates/kernel/src/authority.rs
//! `allows` — the last clause of every command's authorisation.
//!
//! The ruling is that a platform operator has full authority over every party
//! and resource in the deployment. Before this module that was true of eight
//! commands and false of eighteen (`docs/stories/platform-admin-requirements.md`
//! finding **F1**): an operator could revoke a session and could not
//! `set_role`, could transfer a resource and could not `invite`. The gap was
//! not a rule anybody wrote down — it was the absence of one, repeated
//! eighteen times.
//!
//! So there is one function. Every command's authorisation ends in
//! [`allows`], and `platform:* #operator` is its last clause. Not a
//! middleware: the check stays *inside* the command, which is
//! [`crate::cmd`]'s stated rule and the reason a route that forgot to call it
//! would be a hole the type system cannot see. What is shared is the sentence
//! "…or an operator", not the place it is asked.
//!
//! ## The order matters, and it is not an optimisation
//!
//! The ordinary clause is evaluated first and short-circuits. `check()` is the
//! hottest query in the deployment and it is keyed on the subject side; a
//! design that folded the operator relation *into* `check` would add a subject
//! to every expansion and a row to every principal — widening the one query
//! the whole model's latency rests on, for a case that is true for one person
//! in the deployment. The operator relation is therefore a second, separate,
//! indexed read, made only when the ordinary rule already said no.
//!
//! ## Authority is the human's
//!
//! [`allows`] asks about the *person*, never the party a session is acting as
//! ([`crate::cmd::member`]'s middle value). `ActAs` moves attribution; if it
//! moved authority, switching a session would grant something switching it
//! back takes away. The same sentence settles impersonation: an impersonated
//! session's person *is* the target, so it carries the target's authority and
//! not the operator's, and nothing in `check()` learns a new case.

use crate::bind;
use crate::cmd::{Ctx, member};
use crate::error::{KernelError, Outcome, decline};
use crate::feed::Feed;
use crate::ids::{Id, Person};
use crate::org::{self, MemberRole};
use crate::principal::Principal;
use crate::relation::{self, Object, Relation};
use crate::store::{Count, Reads, Sql};

/// What a command needs to be true before it may act.
///
/// Every variant is the command's *ordinary* rule — the one that holds for
/// everybody. [`allows`] adds the operator clause after it, once, in one
/// place.
#[derive(Debug, Clone, Copy)]
pub enum Want {
    /// The command's own rule, already answered. For the handful whose
    /// ordinary rule is not a role or a relation — `may_share`'s two cases,
    /// the last-owner arithmetic, "this identity is mine" — the command works
    /// its own answer out and hands it here, so that even those end in the
    /// same sentence.
    Settled(bool),
    /// A role in a container: an organization or a group, addressed by its
    /// party row.
    Role {
        /// The container's party id.
        container: Id<Person>,
        /// The role the command needs, covered by anything above it.
        role: MemberRole,
    },
    /// A relation on an object, through `check()`.
    On {
        /// What the command needs.
        relation: Relation,
        /// What it needs it on.
        object: Object,
    },
    /// This person and nobody else — the shape a command about somebody's own
    /// row wants. An operator still passes, which is the whole ruling.
    Themselves(Id<Person>),
    /// Nothing but the operator relation. The rules that are not anybody
    /// else's at all: ruling on a merge, registering a relying party,
    /// delegating operator authority.
    Platform,
}

/// May this principal do `want` — as themselves, or as the operator?
///
/// One extra indexed read in the operator case and none in the ordinary one,
/// because the ordinary clause short-circuits.
pub async fn allows<S: Sql, F: Feed>(ctx: &Ctx<'_, S, F>, want: Want) -> Outcome<bool> {
    let (_, person, _) = member(&ctx.principal)?;
    if ordinary(ctx, person, want).await? {
        return Ok(true);
    }
    is_operator(&ctx.store.reads(), person).await
}

/// [`allows`], as the refusal a command returns.
///
/// The uniform decline, because "not permitted" and "no such row" are one
/// answer everywhere else in this crate and an authorisation that said which
/// would be the exception that undoes the rest.
pub async fn require<S: Sql, F: Feed>(ctx: &Ctx<'_, S, F>, want: Want) -> Outcome<()> {
    if allows(ctx, want).await? {
        Ok(())
    } else {
        decline()
    }
}

/// The clause that is not the operator's.
async fn ordinary<S: Sql, F: Feed>(
    ctx: &Ctx<'_, S, F>,
    person: Id<Person>,
    want: Want,
) -> Outcome<bool> {
    match want {
        Want::Settled(held) => Ok(held),
        Want::Role { container, role } => {
            org::holds(&ctx.store.reads(), container, person, role).await
        }
        Want::On { relation, object } => {
            let subjects = crate::principal::expand(&ctx.store.reads(), &ctx.principal).await?;
            relation::check(&ctx.store.reads(), &subjects, relation, object).await
        }
        Want::Themselves(who) => Ok(who == person),
        Want::Platform => Ok(false),
    }
}

/// Whether a person holds `platform:* #operator`.
///
/// Platform administration is a relation, not a column and not a role enum —
/// the kernel report's ruling, and the reason there is no `is_admin` anywhere
/// in this crate. The row is written by the configured bootstrap
/// ([`crate::bootstrap_operator`], which refuses the second one) or by
/// [`crate::cmd::grant_operator`], which has an actor and an audit row.
pub async fn is_operator(store: &impl Reads, person: Id<Person>) -> Outcome<bool> {
    let rows = store.query::<Count>(OPERATOR_SQL, bind![person]).await?;
    Ok(rows.first().is_some_and(|c| c.0 > 0))
}

/// The operator lookup, exposed so the "no full scan" gate can assert that it
/// rides `relation_unique_idx` — it is the last clause of every command in the
/// deployment.
pub const OPERATOR_SQL: &str = "SELECT count(*) AS n FROM relation \
     WHERE object_kind = 'platform' AND object_id = 0 AND relation = 'operator' \
       AND subject_kind = 'person' AND subject_id = $1";

/// How many operators the deployment has. `RevokeOperator`'s last-owner rule.
pub const OPERATOR_COUNT_SQL: &str = "SELECT count(*) AS n FROM relation \
     WHERE object_kind = 'platform' AND object_id = 0 AND relation = 'operator'";

/// How many people operate this deployment.
pub async fn operator_count(store: &impl Reads) -> Outcome<i64> {
    Ok(store
        .query::<Count>(OPERATOR_COUNT_SQL, bind![])
        .await?
        .first()
        .map_or(0, |row| row.0))
}

// ------------------------------------------------------ re-authentication ---

/// How long an operator's session lasts before a platform command asks for the
/// password again. Fifteen minutes: long enough to do a piece of work, short
/// enough that an unattended screen is not a delegation.
pub const PLATFORM_REAUTH_WINDOW: crate::Timestamp = 15 * 60;

/// The commands that decline outside [`PLATFORM_REAUTH_WINDOW`].
///
/// Named rather than derived, because "sensitive" is a judgement about
/// consequences and not a property of the code: what is on this list is what
/// somebody decided a stolen screen must not be able to do. It is also the
/// seam a second factor would gate when passkeys land (requirement A3.3), so
/// it is one list rather than a predicate spread over seven files.
///
/// `disable` is on it for a *person* only. Disabling an organization is
/// undone by enabling it; disabling a person ends every session they hold and
/// stops them signing back in to argue, which is the difference.
pub const SENSITIVE: &[&str] = &[
    "delete-client",
    "disable",
    "disable-product",
    "enable-product",
    "grant-operator",
    "retire-key",
    "revoke-operator",
    "rotate-signing-key",
    "sign-in-as",
];

/// Whether a command name is in the sensitive set.
#[must_use]
pub fn is_sensitive(command: &str) -> bool {
    SENSITIVE.contains(&command)
}

/// When the session behind this principal last presented a password.
///
/// One read by primary key, made only by a sensitive command. It is not on
/// [`Principal`] on purpose: a cached principal would carry a stale
/// `auth_time`, and the one field a re-authentication window must not be stale
/// about is the one it is made of.
const AUTH_TIME_SQL: &str = "SELECT auth_time AS n FROM session WHERE id = $1";

/// Refuse a sensitive command whose session has not seen a password lately.
///
/// The one refusal in this crate a caller is told apart from the others, and
/// the reason is that it is *actionable*. Every other decline is uniform
/// because the difference between "no such thing" and "not yours" is what an
/// attacker is trying to learn; this one tells a caller who is already inside
/// the tier that they must do something they have not been asked to do, and
/// saying nothing would leave them retrying a command that will never work.
/// The server renders it `401`.
///
/// # Errors
///
/// [`KernelError::ReAuthRequired`] when more than [`PLATFORM_REAUTH_WINDOW`]
/// has passed, and the uniform decline when there is no session at all.
pub async fn fresh_enough<S: Sql, F: Feed>(ctx: &Ctx<'_, S, F>) -> Outcome<()> {
    let (_, _, session) = member(&ctx.principal)?;
    let auth_time = ctx
        .store
        .reads()
        .query::<Count>(AUTH_TIME_SQL, bind![session])
        .await?
        .first()
        .map_or(0, |row| row.0);
    if ctx.now() - auth_time > PLATFORM_REAUTH_WINDOW {
        return Err(KernelError::ReAuthRequired);
    }
    Ok(())
}

// ---------------------------------------------------------- impersonation ---

/// What an impersonated session may not do.
///
/// One sentence: *an impersonated session may not change what the person is,
/// or who may become them.* Everything on this list is one of those two, and
/// nothing else is — an operator wearing somebody's hat may write their
/// documents, because that is what the hat is for, and may not add a factor to
/// their account, because a factor outlives the thirty minutes.
pub const FORBIDDEN_WHILE_IMPERSONATING: &[&str] = &[
    "add-factor",
    "disable",
    "grant-operator",
    "re-authenticate",
    "register-client",
    "remove-factor",
    "revoke-operator",
    "rotate-client-secret",
    "set-handle",
    "sign-in-as",
];

/// Whether a command name is refused from an impersonated session.
#[must_use]
pub fn is_forbidden_while_impersonating(command: &str) -> bool {
    FORBIDDEN_WHILE_IMPERSONATING.contains(&command)
}

/// Refuse a command that an impersonated session may not run.
///
/// Asked inside the command rather than in the dispatcher, for the reason
/// every other authorisation in this crate is: a route that forgot to call it
/// would be a hole the type system cannot see, and the ten commands that must
/// refuse are ten places a middleware's list could silently stop matching.
///
/// The uniform decline, unlike [`fresh_enough`]: there is nothing the caller
/// can do about it inside this session, and saying "you are impersonating"
/// back to the one principal that already knows buys nothing.
///
/// # Errors
///
/// The uniform decline when [`Principal::impersonated_by`] is set.
pub fn not_impersonating(principal: &Principal) -> Outcome<()> {
    if principal.impersonated_by().is_some() {
        return decline();
    }
    Ok(())
}

File: src/lib/authority.ts
/* allows(ctx, want). One check, and every feature ends in it.
 *
 * A person is never only themselves. They are also every group they are in,
 * every group that contains one of those, and every organization they belong
 * to — and a relation written against any of those subjects reaches them.
 * `subjectsOf` is that expansion, `allows` is one indexed query over it, and
 * the last clause is always "…or a platform operator" (docs/plan.md §Model).
 *
 * Two vocabularies share one shape. `member < admin < owner` is a role on an
 * organization or a group; `viewer < commenter < editor` is a level on a
 * document, under the document's owner. Nesting is done here, in code: a row
 * saying `editor` answers a question asking for `viewer`, and no command has
 * to remember to look for three verbs.
 *
 * A refusal never says which clause it failed. The caller turns false into
 * the same 404 that a row which does not exist gets, because the difference
 * is exactly the fact a stranger would like to learn. */

import { and, eq, inArray, or, type SQL } from 'drizzle-orm';

import { schema } from '@/db/client';
import type { Transaction } from '@/features/auth/db';

export type SubjectKind = 'person' | 'group' | 'organization';
export interface Subject {
  kind: SubjectKind;
  id: number;
}

export interface Actor {
  personId: number;
  isOperator: boolean;
}

export type Role = 'member' | 'admin' | 'owner';
export type Level = 'viewer' | 'commenter' | 'editor';

export const ROLES: readonly Role[] = ['member', 'admin', 'owner'];
export const LEVELS: readonly Level[] = ['viewer', 'commenter', 'editor'];

const ROLE_RANK: Record<string, number> = { member: 1, admin: 2, owner: 3 };
/* `owner` outranks every level a share can grant: the party a document
 * belongs to is not one of its readers. */
const LEVEL_RANK: Record<string, number> = { viewer: 1, commenter: 2, editor: 3, owner: 4 };

export type ResourceKind = 'organization' | 'group' | 'document';

export interface Want {
  on: ResourceKind;
  id: number;
  need: Role | Level | 'owner';
}

function ranks(on: ResourceKind): Record<string, number> {
  return on === 'document' ? LEVEL_RANK : ROLE_RANK;
}

export function rankOf(on: ResourceKind, verb: string): number {
  return ranks(on)[verb] ?? 0;
}

export function isRole(value: string): value is Role {
  return ROLES.includes(value as Role);
}

export function isLevel(value: string): value is Level {
  return LEVELS.includes(value as Level);
}

/* ---------------------------------------------------------------- subjects */

export interface MembershipEdge {
  subject: Subject;
  /* A group or an organization: the only two things a subject is *in*. */
  resource: Subject;
}

const MAX_DEPTH = 16;

function key(subject: Subject): string {
  return `${subject.kind}:${subject.id}`;
}

/** The transitive set {seed, its groups, the groups containing those, the
 *  organizations it belongs to}. Pure, so nesting and cycles can be stated
 *  without a database — and a cycle simply stops the walk, because a subject
 *  already visited is not visited again. */
export function expandSubjects(seed: Subject, edges: readonly MembershipEdge[]): Subject[] {
  const out: Subject[] = [seed];
  const seen = new Set([key(seed)]);
  let frontier: Subject[] = [seed];

  for (let depth = 0; depth < MAX_DEPTH && frontier.length > 0; depth += 1) {
    const next: Subject[] = [];
    for (const edge of edges) {
      if (!frontier.some((row) => row.kind === edge.subject.kind && row.id === edge.subject.id)) {
        continue;
      }
      if (seen.has(key(edge.resource))) continue;
      seen.add(key(edge.resource));
      out.push(edge.resource);
      next.push(edge.resource);
    }
    frontier = next;
  }
  return out;
}

/** Whether putting `child` inside `parent` would close a loop: it would, iff
 *  the parent is already inside the child. */
export function wouldCycle(
  edges: readonly MembershipEdge[],
  parent: Subject,
  child: Subject,
): boolean {
  if (parent.kind === child.kind && parent.id === child.id) return true;
  return expandSubjects(parent, edges).some(
    (row) => row.kind === child.kind && row.id === child.id,
  );
}

/** Every membership edge in the deployment. The graph is people's circles and
 *  a handful of organizations — small, and read whole so that the walk above
 *  stays one pure function rather than a recursive query nobody can test. */
export async function membershipEdges(tx: Transaction): Promise<MembershipEdge[]> {
  const rows = await tx
    .select({
      subjectKind: schema.relation.subjectKind,
      subjectId: schema.relation.subjectId,
      resourceKind: schema.relation.resourceKind,
      resourceId: schema.relation.resourceId,
    })
    .from(schema.relation)
    .where(
      and(
        inArray(schema.relation.resourceKind, ['group', 'organization']),
        inArray(schema.relation.verb, [...ROLES]),
      ),
    );
  return rows
    .filter((row) => row.subjectKind === 'person' || row.subjectKind === 'group')
    .map((row) => ({
      subject: { kind: row.subjectKind as SubjectKind, id: row.subjectId },
      resource: { kind: row.resourceKind as SubjectKind, id: row.resourceId },
    }));
}

/** Everything a relation could name this person by. */
export async function subjectsOf(tx: Transaction, personId: number): Promise<Subject[]> {
  return expandSubjects({ kind: 'person', id: personId }, await membershipEdges(tx));
}

/* ------------------------------------------------------------------ allows */

/** The `(subject_kind, subject_id) in (…)` half of every relation query. */
export function subjectClause(subjects: readonly Subject[]): SQL | undefined {
  const clauses: SQL[] = [];
  for (const kind of ['person', 'group', 'organization'] as const) {
    const ids = subjects.filter((row) => row.kind === kind).map((row) => row.id);
    if (ids.length === 0) continue;
    clauses.push(
      and(eq(schema.relation.subjectKind, kind), inArray(schema.relation.subjectId, ids))!,
    );
  }
  return clauses.length === 1 ? clauses[0] : or(...clauses);
}

/** The party a resource belongs to, from the registry every command writes.
 *  Organizations are not registered: an organization belongs to nobody but
 *  the people holding `owner` on it. */
async function ownerParty(tx: Transaction, want: Want): Promise<number | null> {
  if (want.on === 'organization') return null;
  const rows = await tx
    .select({ ownerPartyId: schema.resource.ownerPartyId })
    .from(schema.resource)
    .where(and(eq(schema.resource.kind, want.on), eq(schema.resource.refId, want.id)))
    .limit(1);
  return rows[0]?.ownerPartyId ?? null;
}

/** What the database found: who the actor is, who the thing belongs to, and
 *  every verb the two have between them. */
export interface Held {
  subjects: readonly Subject[];
  /* The party the resource belongs to, or null when it belongs to nobody. */
  ownerParty: number | null;
  verbs: readonly string[];
}

/** The decision itself, away from the queries: what a rank is worth, what
 *  ownership is worth, and the operator clause at the end. */
export function decide(held: Held, want: Want, isOperator: boolean): boolean {
  return rankOf(want.on, best(held, want.on) ?? '') >= rankOf(want.on, want.need) || isOperator;
}

/** The best verb the actor holds, or null. The party a thing belongs to holds
 *  every rank over it: an organization that owns a document *is* its owner,
 *  and so is everyone the organization is. Groups are not parties, so a group
 *  subject can never match an owning party by its id. */
export function best(held: Held, on: ResourceKind): string | null {
  if (
    held.ownerParty !== null &&
    held.subjects.some((row) => row.kind !== 'group' && row.id === held.ownerParty)
  ) {
    return 'owner';
  }
  let found: string | null = null;
  for (const verb of held.verbs) {
    if (rankOf(on, verb) === 0) continue;
    if (found === null || rankOf(on, verb) > rankOf(on, found)) found = verb;
  }
  return found;
}

async function heldBy(tx: Transaction, personId: number, want: Want): Promise<Held> {
  const subjects = await subjectsOf(tx, personId);
  const ownerParty_ = await ownerParty(tx, want);
  const where = subjectClause(subjects);
  if (where === undefined) return { subjects, ownerParty: ownerParty_, verbs: [] };
  const rows = await tx
    .select({ verb: schema.relation.verb })
    .from(schema.relation)
    .where(
      and(
        eq(schema.relation.resourceKind, want.on),
        eq(schema.relation.resourceId, want.id),
        where,
      ),
    );
  return { subjects, ownerParty: ownerParty_, verbs: rows.map((row) => row.verb) };
}

export async function allows(tx: Transaction, actor: Actor, want: Want): Promise<boolean> {
  return decide(await heldBy(tx, actor.personId, want), want, actor.isOperator);
}

/** The best verb this actor holds over a resource, for a page that has to
 *  print the word. Null when they hold none — which a page reads as 404. */
export async function heldRank(
  tx: Transaction,
  actor: Actor,
  on: ResourceKind,
  id: number,
): Promise<string | null> {
  const held = await heldBy(tx, actor.personId, { on, id, need: 'owner' });
  const found = best(held, on);
  if (found !== null) return found;
  return actor.isOperator ? 'owner' : null;
}

/** Write a role or level edge. One spelling per subject and resource: a
 *  change of rank replaces the row rather than adding a second one. */
export async function grant(
  tx: Transaction,
  input: { subject: Subject; verb: string; on: ResourceKind; id: number },
): Promise<void> {
  await revoke(tx, { subject: input.subject, on: input.on, id: input.id });
  await tx
    .insert(schema.relation)
    .values({
      subjectKind: input.subject.kind,
      subjectId: input.subject.id,
      verb: input.verb,
      resourceKind: input.on,
      resourceId: input.id,
    })
    .onConflictDoNothing();
}

export async function revoke(
  tx: Transaction,
  input: { subject: Subject; on: ResourceKind; id: number },
): Promise<void> {
  const verbs = input.on === 'document' ? [...LEVELS] : [...ROLES];
  await tx
    .delete(schema.relation)
    .where(
      and(
        eq(schema.relation.subjectKind, input.subject.kind),
        eq(schema.relation.subjectId, input.subject.id),
        inArray(schema.relation.verb, verbs),
        eq(schema.relation.resourceKind, input.on),
        eq(schema.relation.resourceId, input.id),
      ),
    );
}

/** The people who hold `owner` over something directly. What "the last owner
 *  cannot leave" counts. */
export async function ownerPersonIds(
  tx: Transaction,
  on: ResourceKind,
  id: number,
): Promise<number[]> {
  const rows = await tx
    .select({ id: schema.relation.subjectId })
    .from(schema.relation)
    .where(
      and(
        eq(schema.relation.subjectKind, 'person'),
        eq(schema.relation.verb, 'owner'),
        eq(schema.relation.resourceKind, on),
        eq(schema.relation.resourceId, id),
      ),
    );
  return rows.map((row) => row.id);
}

/** Whether removing this person would leave the thing with nobody to own it.
 *  An organization with no owner is an organization nobody can fix. */
export function isLastOwner(owners: readonly number[], personId: number): boolean {
  return owners.includes(personId) && owners.length <= 1;
}

/** Register a resource so that `allows` can find who it belongs to. */
export async function registerResource(
  tx: Transaction,
  input: { kind: ResourceKind | 'event'; id: number; ownerPartyId: number },
): Promise<void> {
  await tx
    .insert(schema.resource)
    .values({ kind: input.kind, refId: input.id, ownerPartyId: input.ownerPartyId })
    .onConflictDoUpdate({
      target: [schema.resource.kind, schema.resource.refId],
      set: { ownerPartyId: input.ownerPartyId },
    });
}
